### PR TITLE
Service Fabric Managed Clusters - initial skd with api version 2021-01-01-preview

### PR DESCRIPTION
**Pull request by a-santamaria**
# All SDK Contribution checklist:

Support for service fabric managed clusters with the api version 2020-01-01-preview swagger link: https://github.com/Azure/azure-rest-api-specs/tree/master/specification/servicefabricmanagedclusters/resource-manager/Microsoft.ServiceFabricManagedClusters/preview/2021-01-01-preview

swagger prs:
initial pr: https://github.com/Azure/azure-rest-api-specs/pull/12639
missing properties and fixes: https://github.com/Azure/azure-rest-api-specs/pull/12858

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [X] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md).**
- [X] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [x] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/master/eng/mgmt/mgmtmetadata).
- [x] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running generate.ps1. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version,

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

### CODE REVIEW

**Comment by a-santamaria** [comment]
adding managed clusters to E:\code\azure-sdk-for-net\eng\pipelines\mgmt.yml here: https://github.com/Azure/azure-sdk-for-net/pull/18888

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+	<!-- Please do not move/edit code below this line -->
+	<Import Project="$(RepoEngPath)/mgmt/AzSdk.reference.props" />
+	<!-- Please do not move/edit code below this line -->
+
+	<PropertyGroup>
+		<PackageId>Microsoft.Azure.Management.ServiceFabricManagedClusters</PackageId>
+		<Description>Microsoft Azure Management ServiceFabricManagedClusters Library</Description>
+		<AssemblyName>Microsoft.Azure.Management.ServiceFabricManagedClusters</AssemblyName>
+		<Version>0.1.0</Version>
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj:10]: First version should be '1.0.0-beta.1'

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+		<PackageTags>Microsoft Azure Management Library;ServiceFabricManagedClusters;ServiceFabricManagedClusters management;</PackageTags>
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj:11]: Also, need to add an 'Properties\AssemblyInfo.cs' file to provide metadata for the assembly (see other projects for specifics)

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/TestHelper/RecordedDelegatingHandler.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+namespace ServiceFabricManagedClusters.Tests.TestHelper
+{
+
+    using System;
+    using System.Net;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Threading.Tasks;
+
+    public class RecordedDelegatingHandler : DelegatingHandler
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/TestHelper/RecordedDelegatingHandler.cs:13]: Why are you adding this?  There is already a similar delegating handler in the test framework

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Tests/ApplicationTests.cs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                Assert.Equal("Succeeded", cluster.ProvisioningState);
+                Assert.Equal("WaitingForNodes", cluster.ClusterState);
+                var primaryNodeType = this.CreateNodeType(serviceFabricMcClient, resourceGroupName, clusterName, nodeTypeName, isPrimary: true, vmInstanceCount: 6);
+
+                this.WaitForClusterReadyState(serviceFabricMcClient, resourceGroupName, clusterName);
+
+                CreateAppType(serviceFabricMcClient, resourceGroupName, clusterName, cluster.ClusterId, AppTypeName);
+                var appTypeVersion = CreateAppTypeVersion(serviceFabricMcClient, resourceGroupName, clusterName, cluster.ClusterId, AppTypeName, AppTypeVersionName, AppPackageUrl);
+                CreateApplication(serviceFabricMcClient, resourceGroupName, clusterName, cluster.ClusterId, AppName, appTypeVersion.Id);
+                CreateService(serviceFabricMcClient, resourceGroupName, clusterName, cluster.ClusterId, AppName, StatefulServiceTypeName, StatefulServiceName);
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Tests/ApplicationTests.cs:49]: There are no assertions over the properties of the returned types.  It is generally a good idea to ensure that the serialization/deserialization of properties is working end-to-end in at east one test

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Generated/Models/ErrorModelException.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+//
+// Code generated by Microsoft (R) AutoRest Code Generator.
+// Changes may cause incorrect behavior and will be lost if the code is
+// regenerated.
+// </auto-generated>
+
+namespace Microsoft.Azure.Management.ServiceFabricManagedClusters.Models
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Generated/Models/ErrorModelException.cs:11]: So, this is occurring because your spec is not referencing the common ARM error model (https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v1/types.json#L277).  Using a custom error like this, rather than the exception form the management library will make it more difficult to handle errors uniformly across RP sdks for the customer

**Review by markcowl — CHANGES_REQUESTED** [review]
Some comments inline.  Also, to make your build and tests run in ci, you need to add a reference to your project here: https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/mgmt.yml

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj (hunk 1 above)
Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj]: changed

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj (hunk 2 above)
Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Microsoft.Azure.Management.ServiceFabricManagedClusters.csproj:11]: added

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/TestHelper/RecordedDelegatingHandler.cs (hunk 3 above)
Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/TestHelper/RecordedDelegatingHandler.cs]: removed

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Tests/ApplicationTests.cs (hunk 4 above)
Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Tests/ApplicationTests.cs:48]: thanks, I added properties validations

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Generated/Models/ErrorModelException.cs (hunk 5 above)
Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/src/Generated/Models/ErrorModelException.cs:11]: Ok. I understand thanks! we will fix this on the next swagger update

**Comment by a-santamaria** [comment]
> Some comments inline. Also, to make your build and tests run in ci, you need to add a reference to your project here: https://github.com/Azure/azure-sdk-for-net/blob/master/eng/pipelines/mgmt.yml

I created a separate pr #18888 as suggested by the check-enforcer comment. Also added it in this pr

**Comment by a-santamaria** [comment]
@markcowl I have addressed the comments, please review the update.

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Microsoft.Azure.Management.ServiceFabricManagedClusters.Test.csproj [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  
+  <ItemGroup>
+    <None Update="SessionRecords\**\*.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
+
+  <!--Do not remove until VS Test Tools fixes #472-->
+  <ItemGroup>
+    <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
```
Comment (markcowl) [inline sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Microsoft.Azure.Management.ServiceFabricManagedClusters.Test.csproj:23]: Why is this needed?

**Review by markcowl — CHANGES_REQUESTED** [review]
One question about the service reference in the test project, otherwise, LGTM

File: sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Microsoft.Azure.Management.ServiceFabricManagedClusters.Test.csproj (hunk 6 above)
Comment (markcowl) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Microsoft.Azure.Management.ServiceFabricManagedClusters.Test.csproj:23]: Please remove this, unless there is a very good reason this is needed

Comment (a-santamaria) [reply sdk/servicefabricmanagedclusters/Microsoft.Azure.Management.ServiceFabricManagedClusters/tests/Microsoft.Azure.Management.ServiceFabricManagedClusters.Test.csproj]: @markcowl removed thanks. I copied it over from somewhere.

**Review by markcowl — APPROVED** [review]
